pkg/lib/machines.js

```diff
@@ -554,15 +555,57 @@
                     state(host, "connecting", null);
             }
 
+            function load_manifest_plugins(packaged_manifests) {
+                if (manifest_plugins[host]) {
+                    manifest_plugins[host].forEach(function (pl) {
+                        pl.close();
+                        $(pl).off();
+                    });
+                }
+                manifest_plugins[host] = [ ];
+
+                function update_manifests() {
+                    var all_manifests = { };
+                    $.extend(true, all_manifests, packaged_manifests);
+                    manifest_plugins[host].forEach(function (pl) {
+                        if (pl.manifests)
+                            $.extend(true, all_manifests, pl.manifests);
+                    });
+                    machines.overlay(host, { manifests: all_manifests });
+                }
+
+                // XXX - This loads and evals all plugins in one go.
+                // Instead, we might want to eval them one by one so
+                // that a crash in one plugin does not disable all
+                // subsequent plugins.
+
+                $.ajax({ url: url + "/*/manifestplugin.js", dataType: "text", cache: true})
```

Comment:
I see.  This would mean that we need to identify and load all components that want to contribute to the menu when the shell is initialized, so that the menu is consistent and doesn't depend on which components have been navigated to by the user.  The components themselves would probably want to know whether they are being used just for the menu additions, or for their full functionality.

I think it is simpler to make dedicated iframes for the manifest plugins.  There not going to be many, I guess.

Also, the current code here assumes that code inside and outside the iframe can freely call each other and exchange JavaScript objects.  This works for same-origin code.  Do you want to restrict this further and force plugins to communicate using postMessage?